workers/unity/Assets/Fps/Scripts/GameLogic/Hud/Mobile/TouchscreenButton.cs

```diff
@@ -0,0 +1,192 @@
+﻿using UnityEngine;
+using UnityEngine.EventSystems;
+using UnityEngine.UI;
+
+[SelectionBase]
+public class TouchscreenButton : MonoBehaviour, IPointerDownHandler, IPointerUpHandler, IDragHandler
+{
+    [Tooltip("How long after pressing the button will register another press")]
+    public float CooldownTime;
+
+    [Tooltip("Whether a click dragged out of the button area ends the press or not")]
+    public bool RestrictPressToButtonArea;
+
+    [Tooltip("The hitbox used to detect if click is dragged out of the button area")]
+    public Graphic Hitbox;
+
+    public bool Togglable;
+
+    public bool IsPressed { get; private set; }
+
+    public delegate void ButtonEvent(PointerEventData eventData);
+
+    public ButtonEvent OnButtonDown;
+    public ButtonEvent OnButtonUp;
+    public ButtonEvent OnButtonDrag;
+
+    private float lastPressedTime;
+
+    private TouchscreenButtonAnimator[] animators;
+
+    private void OnValidate()
+    {
+        CooldownTime = Mathf.Max(0, CooldownTime);
+        OnValidateCheckHitbox();
+    }
+
+    private void OnValidateCheckHitbox()
+    {
+        if (Hitbox != null)
+        {
+            return;
+        }
+
+        var allGraphics = GetComponentsInChildren<Graphic>();
+        foreach (var graphic in allGraphics)
+        {
+            if (!graphic.raycastTarget)
+            {
+                continue;
+            }
+
+            Hitbox = graphic;
+            Debug.LogWarning(
+                $"Automatically located and added Hitbox reference to button {gameObject.name}.\n" +
+                $"You may want to assign your own!\n");
+            break;
+        }
+
+        Debug.LogWarning($"No hitbox found for button {gameObject.name}!\n");
+    }
+
+    private void Awake()
+    {
+        animators = GetComponentsInChildren<TouchscreenButtonAnimator>();
+    }
+
+    // Presently it is assumed the player is dead if the UI is disabled, so reset everything
+    private void OnDisable()
+    {
+        SetPressedTo(false);
+    }
+
+
+    public void SetPressedTo(bool pressed)
+    {
+        // Not terribly nice passing in null PointerEventData refs here,
```

Comment:
> is it preparing for some future usage?

That was the idea. I initially did use PointerEventData for a left-hand fire button, which would need the finger id to tell the move-analogue control to ignore that finger for moving - but I have since removed that fire button as it made the UI more confusing.

We could remove pointerEventData being passed in the events...but it's potentially useful information for future usage. Should we keep it?

Reply:
I'd say remove it unless we're sure it will be used, to keep the code simpler.

Reply:
Removed PointEventData (non)usages, remove SetPressedTo() and replaced with Release() to OnDisable()